oap-server/ai-pipeline/src/main/proto/ai_http_uri_recognition.proto

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+syntax = "proto3";
+
+import "google/protobuf/empty.proto";
+
+
+option java_multiple_files = true;
+option java_package = "org.apache.skywalking.oap.server.ai.pipeline.grpc";
+
+service HttpUriRecognitionService {
+    // Sync for the pattern recognition dictionary.
+    rpc fetchAllPatterns(HttpUriRecognitionSyncRequest) returns (HttpUriRecognitionResponse) {}
+    // Feed new raw data and matched patterns to the AI-server.
+    rpc feedRawData(HttpUriRecognitionRequest) returns (google.protobuf.Empty) {}
+}
+
+message HttpUriRecognitionSyncRequest {
+    string service = 1;
+    // The version of pattern dictionary at the OAP side.
+    string version = 2;
+}
+
+message HttpUriRecognitionRequest {
+    string service = 1;
+    repeated HttpRawUri unrecognizedUris = 2;
+}
+
+message HttpRawUri {
+    string name = 1;
+    int64 matchedCounter = 2;
+}
+
+message HttpUriRecognitionResponse {
+    repeated Pattern patterns = 1;
+    // The version of pattern dictionary at the AI-server side.
+    // If it is as same as HttpUriRecognitionSyncRequest#version,
+    // patterns could be empty as nothing should be updated.
+    string version = 2;
```

Comment:
This version and HttpUriRecognitionSyncRequest#version were my proposed. @Superskyyy I hope this is good enough for you.

Reply:
> This version and HttpUriRecognitionSyncRequest#version were my proposed. @Superskyyy I hope this is good enough for you.

Does it make sense to use `string` as the type of `version`? I'd expect it to be a number so that we can compare (conveniently), tho you might claim we can compare strings too.

Reply:
I just don't want to limit the server. Number(int) is always better but requires comparability, may be too much?

Reply:
> I just don't want to limit the server. Number(int) is always better but requires comparability

Actually I think this is useful in client side, consider this, the client sends `fetchAllPatterns` periodically and, if, for any reasons, the requests are not responded in the same order they are sent, you might update the content to an older version, example:

client send request 1, and 1 minute later, send a request 2, for any reasons, response 2 comes back to client and then later response 1, the client (OAP) will consider response 1 is latest, which is not the case.

This might be a rare case, but I just think it's a low hanging fruit that we can avoid easily by using a number version

Reply:
Is this kind of increasinh version number possible in training?
Especially, remote could be distributed deployment Maintaining this is very hard, remember SkyWalking numberic endpoint ID? Failing bad.

Reply:
> Is this kind of increasing version number possible in training?
> Especially, remote could be distributed deployment Maintaining this is very hard, remember SkyWalking numberic endpoint ID? Failing bad.

Yes, each service can maintain a version number. Whenever a new pattern is discovered or old pattern changes in the batch, the version number will += 1.

Reply:
> Especially, remote could be distributed deployment Maintaining this is very hard, remember SkyWalking numberic endpoint ID? Failing bad.

Are you saying maintaining the version counter in a distributed deployment of the algorithm? If that's the case, the algorithm scales by splitting in units of services, so they will always be only one modifier of each version number concurrently.

Reply:
I am just saying, to keep monotone increasing in a distributed service is not easy. And considering the number of patterns, this seems to not have to limit from protocol perspective.
If we want to make this optimization, r3 could do directly. r3 compares and responds when the version is older rather than different.

Reply:
Anyway, we don't have to do this for now. Let's keep in MVP and make things running.